cellxgene_schema_cli/cellxgene_schema/validate.py

```diff
@@ -1132,21 +1132,46 @@ def _is_raw(self, force: bool = False) -> bool:
             # Get potential raw_X
             raw_loc = self._get_raw_x_loc()
             x = self.adata.raw.X if raw_loc == "raw.X" else self.adata.X
+            if x.dtype != np.float32:
+                self._raw_layer_exists = False
+                self.errors.append("Raw matrix values must have type numpy.float32.")
+                return self._raw_layer_exists
 
             matrix_format = get_matrix_format(self.adata, x)
             assert matrix_format != "unknown"
             self._raw_layer_exists = True
+            has_nonzero_in_every_row = True
+            all_nonzero_valid = True
             for matrix_chunk, _, _ in self._chunk_matrix(x):
-                data = matrix_chunk if isinstance(matrix_chunk, np.ndarray) else matrix_chunk.data
-                if (data % 1 > 0).any():
-                    self._raw_layer_exists = False
+
+                if has_nonzero_in_every_row:
+                    if matrix_format in ("csr", "csc", "coo"):  # full set of sparse matrix types
```

Comment:
- nit: make reusable `("csr", "csc", "coo")` 
- `matrix_format in ("csr", "csc", "coo")` this can be move outside the loop and result used here. Replace 
  `matrix_format = get_matrix_format(self.adata, x)` 
   with 
   `sparse_matrix = get_matrix_format(self.adata, x) in ("csr", "csc", "coo")`